adapters/adkernelAdn/adkernelAdn.go

```diff
@@ -0,0 +1,242 @@
+package adkernelAdn
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+const defaultDomain string = "tag.adkernel.com"
+
+type adkernelAdnAdapter struct{}
+
+//MakeRequests prepares request information for prebid-server core
+func (adapter *adkernelAdnAdapter) MakeRequests(request *openrtb.BidRequest) ([]*adapters.RequestData, []error) {
+	errs := make([]error, 0, len(request.Imp))
+	if len(request.Imp) == 0 {
+		errs = append(errs, newBadInputError("No impression in the bid request"))
+		return nil, errs
+	}
+	imps, impExts, err := getImpressionsInfo(&request.Imp)
+	if len(*imps) == 0 {
+		return nil, *err
+	}
+	errs = append(errs, *err...)
+
+	pub2impressions := dispatchImpressions(imps, impExts)
+	result := make([]*adapters.RequestData, 0, len(pub2impressions))
+	for k, imps := range pub2impressions {
+		bidRequest, err := buildAdapterRequest(request, &k, &imps)
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			result = append(result, bidRequest)
+		}
+	}
+	return result, errs
+}
+
+// getImpressionsInfo checks each impression for validity and returns impressions copy with corresponding exts
+func getImpressionsInfo(imps *[]openrtb.Imp) (*[]openrtb.Imp, *[]openrtb_ext.ExtImpAdkernelAdn, *[]error) {
+	impsCount := len(*imps)
+	errors := make([]error, 0, impsCount)
+	resImps := make([]openrtb.Imp, 0, impsCount)
+	resImpExts := make([]openrtb_ext.ExtImpAdkernelAdn, 0, impsCount)
+
+	for _, imp := range *imps {
+		impExt, err := getImpressionExt(&imp)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		if err := validateImpression(&imp, impExt); err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		resImps = append(resImps, imp)
+		resImpExts = append(resImpExts, *impExt)
+
+	}
+	return &resImps, &resImpExts, &errors
+}
+
+func validateImpression(imp *openrtb.Imp, impExt *openrtb_ext.ExtImpAdkernelAdn) error {
+	if imp.Banner == nil && imp.Video == nil {
```

Comment:
This should now be handled by PR #651 to standardize these basic capability checks, and make sure the advertised capabilities match what the code actually checks for. You can add `glog.Warning("Adkernl CAPABILITY VIOLATION: only banner and video imps supported")` at this time so we can verify that your adapter is never sent and impression that does not allow either of these types. A later PR will pull all these checks from the adapter code to save the adapters a few cycles.

Reply:
Addressed in 076145eba887ea2ac2eab38ffaef77e9a9ef631c